blog/features/replica-signed-queries.md

```diff
@@ -0,0 +1,149 @@
+},
+}
+```
+
+Query response verification is one via agents, such as agent-rs and agent-js, and other agents developed by members of the IC community. Agents now introduce a new `AgentError` type that indicates a verification failure. 
+
+In agent-rs, query response is defined in the file `ic-agent/src/agent/replica_api.rs` as follows:
+
+```
+pub struct CallReply {
+   #[serde(with = "serde_bytes")]
+   pub arg: Vec<u8>,
+   pub signatures: Vec<NodeSignature>,
+}
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
+pub struct RejectResponse {
+    /// The [reject code] returned by the replica.
+    pub reject_code: RejectCode,
+    /// The rejection message.
+    pub reject_message: String,
+    /// The optional [error code]returned by the replica.
+    #[serde(default)]
+    pub error_code: Option<String>,
+    pub signatures: Vec<NodeSignature>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(tag = "status")]
+pub enum QueryResponse {
+   #[serde(rename = "replied")]
+   Replied { reply: CallReply },
+   #[serde(rename = "rejected")]
+   Rejected(RejectResponse),
+}
+```
+
+Replica signed queries are supported in agent-rs [versions v0.30.0](https://github.com/dfinity/agent-rs/releases) and newer.
+
+In agent-js, query response is defined in the file `agent-js/packages/agent/src/agent/api.ts` as follows:
+
+```
+export type QueryResponse = QueryResponseReplied | QueryResponseRejected;
+	
+	export const enum QueryResponseStatus {
+	 Replied = 'replied',
+	 Rejected = 'rejected',
+	}
+	
+	export interface QueryResponseBase {
+	 	status: QueryResponseStatus;
+               signatures: Array<NodeSignature>;
+	}
+	
+	export interface QueryResponseReplied extends QueryResponseBase {
+	 status: QueryResponseStatus.Replied;
+	 reply: { arg: ArrayBuffer };
+	}
+	
+	export interface QueryResponseRejected extends QueryResponseBase {
+	 status: QueryResponseStatus.Rejected;
+	 reject_code: ReplicaRejectCode;
+	 reject_message: string;
+	}
+```
+
+Replica signed queries are supported in agent-js [versions v0.20.0](https://github.com/dfinity/agent-js/releases) and newer.
+
+## How are query responses verified? 
+
+To verify a replica signed query, first the agent requests the path `/subnet` in a read state call that is separate to `/api/v2/canister/<effective_canister_id>/read_state` to get the certificate. A boundary node routes this call to the subnet that hosts the canister. A subnet's state tree only contains the keys of its own nodes. All node keys are cached.
+
+Then, to verify the certificate, the timestamp stored at `/time` of the node keys will be checked to assure that the nodes are still within the subnet. The agent can look up the key of the node that is present in the query response signature, either from the cache or the `read_state` response. Once the node's public key is obtained, the client can verify the signature against the hash of the response's relevant fields. 
+
+## Conclusion
```

Comment:
Before the conclusion, I am thinking if we can add a section about how users can start benefiting from the feature, i.e. when the feature becomes effective. There are three aspects.
1. The replica has already producing signatures on query responses. Old version of agents automatically ignores the signature field.
2. For agent-rs, signature verification on query responses is by default enabled on [versions v0.30.0] and newer. Tools like dfx will also start using new version of agent-rs in new releases.
3. For agent-js,
(1) Already existing applications will have to update the agent-js version to [versions v0.20.0] and newer.
(2) Dfx will always use the agent-js version tagged as @latest when running dfx new. Therefore, new applications will automatically enable signature verification on query responses.

@krpeacock @adamspofford-dfinity Could you help take a look at these statements and offer some advice? Thank you.